sdk/python/lib/pulumi/runtime/rpc_manager.py

```diff
@@ -60,49 +59,28 @@ def do_rpc(self, name: str, rpc_function: Callable[..., Awaitable[Tuple[Any, Exc
         """
         async def rpc_wrapper(*args, **kwargs):
             log.debug(f"beginning rpc {name}")
-            async with self.zero_cond:
-                self.count += 1
-                log.debug(f"recorded new RPC, {self.count} RPCs outstanding")
+            self.count += 1
+            log.debug(f"recorded new RPC, {self.count} RPCs outstanding")
 
             try:
                 result = await rpc_function(*args, **kwargs)
                 exception = None
             except Exception as exn:
                 log.debug(f"RPC failed with exception:")
                 log.debug(traceback.format_exc())
-                if not self.exception_future.done():
-                    self.exception_future.set_exception(exn)
+                if self.unhandled_exception is None:
+                    self.unhandled_exception = exn
+                    self.exception_traceback = sys.exc_info()[2]
```

Comment:
Is it worth logging or otherwise handling the case that we get an exception here and `unhandled_exception` is not None?

It feels like any case where that happens will lead to at least slightly confusing behaviour for users.

Reply:
We can log in this case, yes.

Reply:
On second thought, we already log on every exception that happens in the context of `do_rpc`, so I don't think we need the extra logging here.